src/main/java/org/broadinstitute/hellbender/tools/copynumber/datacollection/PerBaseCountCollector.java

```diff
@@ -0,0 +1,150 @@
+package org.broadinstitute.hellbender.tools.copynumber.datacollection;
+
+import htsjdk.samtools.util.Locatable;
+import org.broadinstitute.hellbender.tools.copynumber.formats.collections.PerBaseCountCollection;
+import org.broadinstitute.hellbender.tools.copynumber.formats.metadata.SampleLocatableMetadata;
+import org.broadinstitute.hellbender.tools.copynumber.formats.records.PerBaseCount;
+import org.broadinstitute.hellbender.utils.Nucleotide;
+import org.broadinstitute.hellbender.utils.SimpleInterval;
+import org.broadinstitute.hellbender.utils.Utils;
+import org.broadinstitute.hellbender.utils.param.ParamUtils;
+import org.broadinstitute.hellbender.utils.pileup.PileupElement;
+import org.broadinstitute.hellbender.utils.pileup.ReadPileup;
+
+import java.util.*;
+
+/**
+ * Collects per-base counts at specified sites where bases are one of those specified in {@link PerBaseCountCollector#BASES}.
+ * If two reads share the same query name and, therefore, originate from the same template of DNA, this
+ * only counts the higher base quality nucleotide of the two reads. If the base qualities of the two reads are equal,
+ * the nucleotide from read one will be counted.
+ *
+ * @author Samuel Lee &lt;slee@broadinstitute.org&gt;
+ * @author Robert Klein &lt;rklein@broadinstitute.org&gt;
+ */
+public class PerBaseCountCollector {
+    public static final List<Nucleotide> BASES = Collections.unmodifiableList(Arrays.asList(Nucleotide.A, Nucleotide.C, Nucleotide.G, Nucleotide.T, Nucleotide.N));
+
+    private final SampleLocatableMetadata metadata;
+    private final List<PerBaseCount> perBaseCounts = new ArrayList<>();
+
+    public PerBaseCountCollector(final SampleLocatableMetadata metadata) {
+        this.metadata = Utils.nonNull(metadata);
+    }
+
+    /**
+     * Add counts to this class for a specific locus.
+     *
+     * @param pileup associated pileup at the locus.  Not {@code null}
+     * @param locus position in genome to collect alellic counts.  Not {@code null}
+     * @param minBaseQuality minimum base quality in the read for that read to count at that position.  Must be greater than or equal to 0.
+     */
+    public void collectAtLocus(final ReadPileup pileup, final Locatable locus, final int minBaseQuality) {
+        Utils.nonNull(pileup);
+        Utils.nonNull(locus);
+        ParamUtils.isPositiveOrZero(minBaseQuality, "Minimum base quality must be zero or higher.");
+
+        final HashMap<Nucleotide, Integer> perBaseCount = new HashMap<>();
+        for(Nucleotide base : BASES) {
+            perBaseCount.put(base, 0);
+        }
+
+        // Track the base quality of the first encountered read of a pair until the pair is found or iteration ends
```

Comment:
Can we make this logic more concise with the appropriate data structure?  The nested if-else blocks and the repeated code block for adding counts seem like a code smell to me.

Perhaps a multimap keyed on the query name (with appropriate handling of null), with the collection of values storing the necessary data (first-of-pair, BQ, base) to determine which base call to keep for each read?  I think it might be cleaner to dump the entire pileup into this data structure, then iterate through and apply the appropriate logic/comparator to pull out the desired base call---probably could do it with two stream statements.